packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx

```diff
@@ -320,6 +320,56 @@ describe('<DataGrid /> - Pagination', () => {
         );
         expect(getColumnValues(0)).to.deep.equal(['7', '8']);
       });
+
+      it.only('should update the amount of rows rendered and call onPageSizeChange when changing the table height', async () => {
+        const onPageSizeChange = spy();
+
+        const TestCaseAutoPageSize = (props: { nbRows: number; height?: number }) => {
+          const data = useData(props.nbRows, 10);
+
+          return (
+            <div style={{ width: 300, height: props.height }}>
+              <DataGrid
+                columns={data.columns}
+                rows={data.rows}
+                autoPageSize
+                pagination
+                onPageSizeChange={onPageSizeChange}
+              />
+            </div>
+          );
+        };
+        const nbRows = 27;
+        const heightBefore = 780;
+        const heightAfter = 360;
+
+        const { setProps } = render(<TestCaseAutoPageSize nbRows={nbRows} height={heightBefore} />);
+
+        const footerHeight = document.querySelector('.MuiDataGrid-footerContainer')!.clientHeight;
+        const expectedViewportRowsLengthBefore = Math.floor(
+          (heightBefore - DEFAULT_GRID_OPTIONS.headerHeight - footerHeight) /
+            DEFAULT_GRID_OPTIONS.rowHeight,
+        );
+        const expectedViewportRowsLengthAfter = Math.floor(
+          (heightAfter - DEFAULT_GRID_OPTIONS.headerHeight - footerHeight) /
+            DEFAULT_GRID_OPTIONS.rowHeight,
+        );
+
+        let rows = document.querySelectorAll('.MuiDataGrid-viewport [role="row"]');
+        expect(rows.length).to.equal(expectedViewportRowsLengthBefore);
+
+        setProps({ height: heightAfter });
+
+        await raf();
+        await sleep(1000);
```

Comment:
@dtassone  did you have trouble with Safari when testing something that requires a redraw of the table ?
If I just wait for the next animation frame, it works on Chrome and Edge but Safari does not yet have the new value (the useEffect that publish the events is called after the await raf() finishes.
But if I wait for an additional second it does work.

Reply:
```
Chrome Headless 93.0.4530.0 (Linux x86_64) WARN: 'including inaccessible elements by default'
LOG: 'CODE: After line "apiRef.current.publishEvent(GRID_PAGE_SIZE_CHANGE, { pageSize: autoPageSize })"', Object{prevPageSize: 100, autoPageSize: 12}
LOG: 'TEST: after line "setProps({ height: heightAfter })"'
LOG: 'CODE: After line "apiRef.current.publishEvent(GRID_PAGE_SIZE_CHANGE, { pageSize: autoPageSize })"', Object{prevPageSize: 12, autoPageSize: 4}
LOG: 'TEST: after line "await raf()"'
LOG: 'TEST: after line "await sleep(100)"'
```

```
Safari 13.1.2 (Mac OS 10.15.6) LOG: 'CODE: After line "apiRef.current.publishEvent(GRID_PAGE_SIZE_CHANGE, { pageSize: autoPageSize })"', Object{prevPageSize: 100, autoPageSize: 12}
Safari 13.1.2 (Mac OS 10.15.6) LOG: 'TEST: after line "setProps({ height: heightAfter })"'
Safari 13.1.2 (Mac OS 10.15.6) LOG: 'TEST: after line "await raf()"'
Safari 13.1.2 (Mac OS 10.15.6) LOG: 'CODE: After line "apiRef.current.publishEvent(GRID_PAGE_SIZE_CHANGE, { pageSize: autoPageSize })"', Object{prevPageSize: 12, autoPageSize: 4}
Safari 13.1.2 (Mac OS 10.15.6) LOG: 'TEST: after line "await sleep(100)"'
```

The execution order in not the same, Safari triggers the `React.useEffect` later

Reply:
Note that you can debug BrowserStack on your local env with https://trello.com/c/KecHQ7CB/2773-browserstack.

Reply:
I run BrowserStack in my local env yesterday and it does work
But the callback is triggered a bit later so the test fails if I don't put a `sleep(100)` in it, which is probably not the best solution.
I guess that there is a better option on the test because it's seems like a very common situation to wait for a rerender caused by a state update in a `useEffect`.

Reply:
https://github.com/mui-org/material-ui-x/pull/1986/commits/b8631524dcc1fe82d48728345e851e53e184695d

I have a workaround, I can just wait for the table to actually have the expected amount of rows and then check if the callback has been called.